_deploy/r/gnoswap/gns/gns.gno

```diff
@@ -93,93 +129,112 @@ func Render(path string) string {
-func Mint(address pusers.AddressOrName) uint64 {
-	common.IsHalted()
-
-	caller := std.PrevRealm().Addr()
-	if caller != consts.EMISSION_ADDR {
-		panic(addDetailToError(
-			errNoPermission,
-			ufmt.Sprintf("only emission contract(%s) can call Mint, called from %s", consts.EMISSION_ADDR, caller.String()),
-		))
+func checkErr(err error) {
+	if err != nil {
+		panic(err.Error())
 	}
+}
 
-	// if not yet initialized, mint 0 amount
-	if initialized == false {
+// calculateAmountToMint calculates the amount of gns to mint
+// It calculates the amount of gns to mint for each halving year for block range.
+// It also handles the left emission amount if the current block range includes halving year end block.
+func calculateAmountToMint(fromHeight, toHeight int64) uint64 {
+	fromYear := GetHalvingYearByHeight(fromHeight)
+	toYear := GetHalvingYearByHeight(toHeight)
+
+	if isEmissionEnded(fromYear) || isEmissionEnded(toYear) {
 		return 0
 	}
 
-	// calculate gns emission amount for every block, and send by single call
-	// for this case, we assume that inside of block range gnoswap state hasn't changed.
-	nowHeight := std.GetHeight()
-	amountToMint := uint64(0)
+	totalAmountToMint := uint64(0)
 
-	if lastMintedHeight >= nowHeight {
-		return 0
-	}
+	for i := fromYear; i <= toYear; i++ {
+		yearEndHeight := GetHalvingYearBlock(i)
+		mintUntilHeight := yearEndHeight
 
-	// If from, to block is at same halving year, no need iterate
-	fromYear := GetHalvingYearByHeight(lastMintedHeight + 1)
-	toYear := GetHalvingYearByHeight(nowHeight)
-
-	if fromYear == toYear {
-		numBlock := uint64(nowHeight - lastMintedHeight)
-		singleBlockAmount := GetAmountByHeight(nowHeight)
-		totalBlockAmount := singleBlockAmount * numBlock
-
-		amountToMint += totalBlockAmount
-		amountToMint = checkAndHandleIfLastBlockOfHalvingYear(nowHeight, amountToMint)
-
-		halvingYearMintAmount[fromYear] += totalBlockAmount
-	} else {
-		for i := lastMintedHeight + 1; i <= nowHeight; i++ {
-			amount := GetAmountByHeight(i)
-			amount = checkAndHandleIfLastBlockOfHalvingYear(i, amount)
-			year := GetHalvingYearByHeight(i)
-			halvingYearMintAmount[year] += amount
-			amountToMint += amount
+		if toHeight < mintUntilHeight {
+			mintUntilHeight = toHeight
 		}
-	}
 
-	err := privateLedger.Mint(users.Resolve(address), amountToMint)
-	if err != nil {
-		panic(err.Error())
+		// how many blocks to calculate
+		numBlock := uint64(mintUntilHeight-fromHeight) + 1
+
+		// amount of gns to mint for each block for current year
+		singleBlockAmount := GetAmountByHeight(yearEndHeight)
+
+		// amount of gns to mint for current year
+		yearAmountToMint := singleBlockAmount * numBlock
+
+		// if last block of halving year, handle left emission amount
+		if isLastBlockOfHalvingYear(mintUntilHeight) {
+			yearAmountToMint += handleLeftEmissionAmount(i, yearAmountToMint)
+			totalAmountToMint += yearAmountToMint
+		} else {
+			totalAmountToMint += yearAmountToMint
+		}
```

Comment:
```suggestion
			yearAmountToMint += handleLeftEmissionAmount(i, yearAmountToMint)
		}
		totalAmountToMint += yearAmountToMint
```

Regardless of the `isLastBlockOfHalvingYear` condition, `yearAmountToMint` is always being added to `totalAmountToMint`. In this case `else` block is unnecessary.